persisted queries: improve safelisting semantics

glasser opened the pull request:
This refactor does a few things:

- When the safelist is enabled (but `require_id` is not), we used to require the incoming freeform GraphQL operation to be a precise byte-for-byte match for an operation in the manifest. However, this can be challenging to accomplish. For example, Apollo Client Web projects can be laid out in many ways and the way that fragment definitions are connected to operation definitions is often challenging for tooling to accurately predict at build time. The order of fragment and operation definitions in a document has no semantic effect, so in order to make the job of the build-time operation extraction tool easier, Router now sorts top-level definitions deterministically when constructing a safelist and checking incoming operations against it. This also normalizes ignored tokens (whitespace, comments, commas, etc) so that differences in these tokens do not affect safelisting.

- Previously, we had a hard-coded hack to allow two particular introspection queries sent by a particular version of Apollo Sandbox even when in safelisting mode (even with require_id:true). This PR uses `apollo-rs` to more accurately detect pure-introspection queries, and so any query that only queries the top-level introspection fields will be allowed to pass the safelist. However, this PR also changes the behavior of `require_id: true` to not have this exception, so that `require_id: true` will never parse any incoming freeform GraphQL requests (even just to determine if they are introspection). (These servers can still run introspection requests if the introspection query is registered and run by ID!)

- Previously, when dynamically reloading configuration, the PersistedQueryManifestPoller object from the previous iterated would be preserved rather than restarted. This made config reloads faster when PQs are enabled. However, it meant that the PersistedQueryManifestPoller could not be sensitive to the YAML configuration file, because it was shared across multiple reloads. Additionally, the mechanism used to share the object was buggy: if the feature was disabled, the `new_disabled` function would shut down the background poll routine for the *shared* poller object, and if for some reason the reload failed for another reason after that point, it would never be restarted! It would certainly be possible to refactor the poller so it could be shared correctly (with the shutdown occurring only when the poller is fully dropped) and where it could notify its surrounding PQ layers via a broadcast channel or something when the state changes so that configuration-sensitive actions could be taken. This PR takes a simpler approach, of not trying to preserve the poller across reloads (similar to the JwksManager in the authentication plugin). Now the poller object can update its internal state with a safelist when configuration says that a safelist is necessary.

- Previously, if `log_unknown: true` ("audit mode") and APQs were both enabled, we would *not* log requests that contain APQ IDs as unknown, even if the operation in question is not in the manifest. This meant that the "audit mode" wouldn't actually tell you about all requests that would fail if safelisting were enabled (which does require disabling APQs)! After this PR, the `log_unknown: true` message is logged whenever an operation not in the manifest is executed, even if it is loaded from the APQ cache.

- Previously, if `log_unknown: true` was combined with `safelist.require_id: true`, Router would only log "unknown operation" for operations not in the safelist: it would not log anything when rejecting a "safelisted" operation that is sent as freeform GraphQL rather than by ID. After this PR, Router logs "unknown operation" for all freeform GraphQL operations in this configuration. (It no longer differentiates between those that would have been accepted if the safelist were enabled without requiring ID and those that would have been rejected in that case; this means that the Router does not have to generate a normalized safelist at all when `safelist.require_id: true` is on.)

glasser commented:
No doc updates because the docs for this feature are only on `main`, not `dev`. I'll make a changelog entry once this has been reviewed a bit more.

o0Ignition0o commented:
> These servers can still run introspection requests if the introspection query is registered and run by ID!

Does this have a studio ux impact? IE going to explorer showing operation collections and sending queries by ID instead of displaying a (broken) sandbox?

o0Ignition0o commented on apollo-router/src/services/layers/persisted_queries/mod.rs:
```diff
@@ -184,57 +150,84 @@ impl PersistedQueryLayer {
+                None => {
+                    drop(context_guard);
+                    // For some reason, QueryAnalysisLayer didn't give us a Compiler?
+                    return Err(supergraph_err(
+                        graphql_err(
+                            "MISSING_PARSED_OPERATION",
+                            "internal error: compiler missing",
+                        ),
+                        request,
+                        ErrorCacheStrategy::DontCache,
+                    ));
```
We tend to rely on https://doc.rust-lang.org/std/ops/enum.ControlFlow.html to differenciate between errors that will yield an http response, (ControlFlow::Break) and BoxErrors (HTTP 500).

This looks like this should be a BoxError. Did we consider using Result<ControlFlow<SupergraphRequest, SupergraphResponse>, BoxError>> ?

glasser replied:
I'm not too familiar with that enum, and it's not currently used anywhere in this layer or its friends like the APQ layer or query analysis layer. Is the idea that I would make the top level entry points to the layer (`supergraph_request` etc) return this enum and then update the code in `router_service.rs` to process its output?

o0Ignition0o replied:
I see, that's ok then. I tend to rely on controlflow when i need to differenciate between errors I want to reply to the client, and router errors that will bubble into a 500.

`compiler missing` looks like something a client should never see.

EverlastingBugstopper replied:
ah that checks out. probably worth doing then! what do you think about failure mode for being unable to acquire locks? currently it's a panic.

glasser replied:
FWIW I modeled this after BridgeQueryCompiler::call which does what looks like the same thing as a parse error if this object is missing... Not sure what change to make here.

o0Ignition0o replied:
> what do you think about failure mode for being unable to acquire locks? currently it's a panic.
I think this is one of the rare cases where a panic would be good. I'm not sure how to recover from a poisoned lock.

> FWIW I modeled this after BridgeQueryCompiler::call which does what looks like the same thing as a parse error if this object is missing... Not sure what change to make here.

I see, then there's probably a bit of work for us to do in both layers, would you mind opening a followup issue?

glasser commented on apollo-router/src/services/layers/persisted_queries/mod.rs:
```diff
@@ -1,121 +1,70 @@
+        if configuration.preview_persisted_queries.safelist.require_id
+            && !configuration.preview_persisted_queries.safelist.enabled
+        {
+            return Err(anyhow!("invalid configuration: preview_persisted_queries.safelist.require_id = true requires you to also set preview_persisted_queries.safelist.enabled = true").into());
```
Is your suggestion to just not use `anyhow` at all anywhere, or to only use it when there's some non-trivial formatting and stick to `"foo".into()` for the purely-a-string-constant case?

o0Ignition0o replied:
AIUI there's an impl From<&str> and an impl From<String> for boxerror. You could replace non trivial anyhows with format!("your error: {}", e).into().

Note it's more of a nit than a blocker so it's up to you.

glasser commented on apollo-router/src/services/layers/persisted_queries/mod.rs:
```diff
@@ -1,121 +1,70 @@
-                    return Err(anyhow!("invalid configuration: preview_persisted_queries.safelist.enabled = true, which is incompatible with apq.enabled = true. you must disable apq in your configuration to enable persisted queries with safelisting").into());
-                }
-                Self::new_enabled(configuration, uplink_config, previous_manifest_poller).await
+            if configuration.apq.enabled && configuration.preview_persisted_queries.safelist.enabled
+            {
+                Err(anyhow!("invalid configuration: preview_persisted_queries.safelist.enabled = true, which is incompatible with apq.enabled = true. you must disable apq in your configuration to enable persisted queries with safelisting").into())
             } else {
-                Err(anyhow!("persisted queries requires Apollo GraphOS. ensure that you have set APOLLO_KEY and APOLLO_GRAPH_REF environment variables").into())
+                Ok(Self {
+                    manifest_poller: Some(
+                        PersistedQueryManifestPoller::new(configuration.clone()).await?,
+                    ),
+                })
             }
+        } else if configuration.preview_persisted_queries.safelist.enabled {
+            Err(anyhow!("invalid configuration: preview_persisted_queries.safelist.enabled = true, which is incompatible with preview_persisted_queries.enabled = false. you must enabled persisted queries to enable safelisting").into())
+        } else if configuration.preview_persisted_queries.log_unknown {
+            Err(anyhow!("invalid configuration: preview_persisted_queries.log_unknown = true, which is incompatible with preview_persisted_queries.enabled = false. you must enabled persisted queries to enable log_unknown").into())
         } else {
```
That works for me. Actually I was wondering if there was a different pattern for how to do basic YAML structure validation like "if you don't pass `enabled:true` you don't get to do the other parts of the feature" — am I right to put that in this file at all or does it go somewhere in the `configuration` directory?

glasser replied:
I see there's already a Configuration::validate function now that I check: I'll put it in there (or well, I'll put it in the persisted_queries.rs file as a function called from there?)

glasser commented:
> > These servers can still run introspection requests if the introspection query is registered and run by ID!
> Does this have a studio ux impact? IE going to explorer showing operation collections and sending queries by ID instead of displaying a (broken) sandbox?

The only impact of this particular change is that if you are running with `preview_persisted_queries: safelist: require_id: true` and `supergraph: introspection: true` (honestly an unlikely combination) and you go to *Sandbox* (not non-Sandbox Explorer, which uses the schema registry rather than introspection) you won't be able to see the schema. But either way you won't be able to execute operations afterwards.  (We do have some plans to add an optional feature to improve Explorer's interaction with safelisting in different ways.)

EverlastingBugstopper commented on apollo-router/src/services/layers/persisted_queries/mod.rs:
```diff
@@ -1,121 +1,70 @@
 use crate::Configuration;
-use crate::UplinkConfig;
 
-const SANDBOX_INTROSPECTION_QUERY_WO_DEPRECATED_ARGS: &str = "\n    query IntrospectionQuery {\n      __schema {\n        \n        queryType { name }\n        mutationType { name }\n        subscriptionType { name }\n        types {\n          ...FullType\n        }\n        directives {\n          name\n          description\n          \n          locations\n          args {\n            ...Inp ...
-const SANDBOX_INTROSPECTION_QUERY_W_DEPRECATED_ARGS: &str = "\n    query IntrospectionQuery {\n      __schema {\n        \n        queryType { name }\n        mutationType { name }\n        subscriptionType { name }\n        types {\n          ...FullType\n        }\n        directives {\n          name\n          description\n          \n          locations\n          args(includeDeprecated: tru ...
```
😍

EverlastingBugstopper commented on apollo-router/src/services/layers/persisted_queries/mod.rs:
```diff
@@ -184,57 +150,84 @@ impl PersistedQueryLayer {
             }
+        };
+
+        // Introspection requests are always allowed. (This means any document
```
should we skip this step if introspection is disabled?

EverlastingBugstopper commented on apollo-router/src/services/layers/persisted_queries/manifest_poller.rs:
```diff
@@ -18,102 +22,303 @@ use crate::uplink::persisted_queries_manifest_stream::PersistedQueriesManifestCh
+
+    fn insert_from_manifest(&mut self, body_from_manifest: &str) {
+        self.normalized_bodies.insert(
+            self.normalize_body(body_from_manifest, Parser::new(body_from_manifest).parse()),
```
is this a semi-permanent decision or one made in pursuit of rapid development? if the latter, let's make a comment to re-visit pulling a parser out of the compiler that's in context from the `query_analysis_layer`.

glasser replied:
@EverlastingBugstopper I'm not sure I follow. This is the function that runs when you load a manifest, not on an incoming request.

EverlastingBugstopper replied:
oh shoot you're totally right, disregard

EverlastingBugstopper reviewed (approved):
looking great!! left a few non-blocking comments

EverlastingBugstopper reviewed (approved):
looking great!! left a few non-blocking comments

glasser commented:
OK, I think I've addressed the feedback in these latest commits. I also realized I had accidentally made `log_unknown` not log rejections (only "would be rejections if safelisting was on") so I fixed that too.

o0Ignition0o commented on apollo-router/src/configuration/mod.rs:
```diff
@@ -437,6 +438,36 @@ impl Configuration {
             );
         }
 
+        // PQs.
+        if self.preview_persisted_queries.enabled {
+            if self.preview_persisted_queries.safelist.enabled && self.apq.enabled {
+                return Err(ConfigurationError::InvalidConfiguration {
+                    message: "apqs must be disabled to enable safelisting",
+                    error: "either set preview_persisted_queries.safelist.enabled = false or apq.enabled = false".into()
+                });
+            } else if !self.preview_persisted_queries.safelist.enabled
+                && self.preview_persisted_queries.safelist.require_id
+            {
+                return Err(ConfigurationError::InvalidConfiguration {
+                    message: "safelist must be enabled to require IDs",
+                    error: "either set preview_persisted_queries.safelist.enabled = true or preview_persisted_queries.safelist.require_id = false".into()
+                });
+            }
+        } else {
+            // If the feature isn't enabled, sub-features shouldn't be.
+            if self.preview_persisted_queries.safelist.enabled {
+                return Err(ConfigurationError::InvalidConfiguration {
+                    message: "persisted queries must be enabled to enable safelisting",
+                    error: "either set preview_persisted_queries.safelist.enabled = false or preview_persisted_queries.enabled = true".into()
+                });
+            } else if self.preview_persisted_queries.log_unknown {
+                return Err(ConfigurationError::InvalidConfiguration {
+                    message: "persisted queries must be enabled to enable logging unknown operations",
+                    error: "either set preview_persisted_queries.log_unknown = false or preview_persisted_queries.enabled = true".into()
+                });
+            }
+        }
+
```
love it!

o0Ignition0o commented on apollo-router/src/services/layers/persisted_queries/mod.rs:
```diff
@@ -184,61 +143,98 @@ impl PersistedQueryLayer {
+            Some(mp) => mp,
+        };
+
+        let operation_body = match request.supergraph_request.body().query.clone() {
+            // if the request doesn't have a `query` document, continue with normal execution, which
+            // will result in the normal no-operation error.
+            None => return Ok(request),
+            Some(ob) => ob,
+        };
```
nit: I think you can get away with a reference here (and avoid cloning) using [as_ref](https://doc.rust-lang.org/std/option/enum.Option.html#method.as_ref)

```suggestion
        let operation_body = match request.supergraph_request.body().query.as_ref() {
            // if the request doesn't have a `query` document, continue with normal execution, which
            // will result in the normal no-operation error.
            None => return Ok(request),
            Some(ob) => ob,
        };
```

You would then pass operation_body instead of &operation_body below.

glasser replied:
Done!

o0Ignition0o reviewed (approved):
Looking great 🚀 
I've added a tiny nit but honestly it won't have a huge impact imo.

LGTM

glasser commented:
It actually turns out that I don't need to do a docs change as part of this PR, because the relevant parts of the docs are in shared components stored in the `docs` repo.

I will need to do CHANGELOG though.

glasser commented:
I added a changeset.

EverlastingBugstopper reviewed (approved):
LGTM! 🚢

BrynCooke commented on apollo-router/src/axum_factory/tests.rs:
```diff
@@ -2320,7 +2320,6 @@ async fn test_supergraph_timeout() {
         QueryAnalysisLayer::new(supergraph_creator.schema(), Arc::clone(&conf)).await,
         Arc::new(supergraph_creator),
         conf.clone(),
-        Default::default(),
```
Super pleased that the manifest poller is no-longer leaking out of the layer.

glasser replied:
Agreed, though there is a downside there — a pattern for (correctly!) preserving state over config reloads would be interesting.

glasser commented on apollo-router/src/configuration/mod.rs:
```diff
@@ -437,6 +438,36 @@ impl Configuration {
             );
         }
 
+        // PQs.
```
Agreed on both parts.  (My instinct would be that this would be something like allow_clients_to_register_operations to be concrete instead of marketing-y but probably that's not great given how long we've had that name.) Perhaps we do this as part of removing `preview_`.

glasser commented on apollo-router/src/configuration/mod.rs:
```diff
@@ -437,6 +438,36 @@ impl Configuration {
+            if self.preview_persisted_queries.safelist.enabled && self.apq.enabled {
+                return Err(ConfigurationError::InvalidConfiguration {
+                    message: "apqs must be disabled to enable safelisting",
+                    error: "either set preview_persisted_queries.safelist.enabled = false or apq.enabled = false".into()
```
Makes sense and done, although I'm not sure that the dot syntax is really YAMLy either.

EverlastingBugstopper commented on .changesets/exp_pq_safelist_normalization.md:
```diff
@@ -7,6 +7,7 @@ Several improvements to safelisting behavior based on preview feedback:
 * When the safelist is enabled (but `require_id` is not), matching now ignores the order of top-level definitions (operations and fragments) and ignored tokens (whitespace, comments, commas, etc), so that differences in these purely syntactic elements do not affect whether an operation is considered to be in the safelist.
 * If introspection is enabled on the server, any operation whose top-level fields are introspection fields (`__type`, `__schema`, or `__typename`) is considered to be in the safelist. (Previously, Router instead looked for two specific introspection queries from a particular version of Apollo Sandbox if sandbox was enabled; this hard-coded check is removed.) This special case is not applied if `r ...
 * When `log_unknown` is enabled and `apq` has not been disabled, Router now logs any operation not in the safelist as unknown, even those sent via IDs if the operation was found in the APQ cache rather than the manifest.
+* When `log_unknown` and `require_id` are both enabled, Router now logs all operations that rejects (i.e., all operations sent as freeform GraphQL). Previously, Router only logged the operations that would have been rejected by the safelist feature with `required_id` disabled (i.e., operations sent as freeform GraphQL that do not match an operation in the manifest).
```
```suggestion
* When `log_unknown` and `require_id` are both enabled, Router now logs all operations that rejects (i.e., all operations sent as freeform GraphQL). Previously, Router only logged the operations that would have been rejected by the safelist feature with `require_id` disabled (i.e., operations sent as freeform GraphQL that do not match an operation in the manifest).
```

glasser commented:
I added another line to the changelog and PR description describing a change from this PR that was unintentional but I consider positive. (Some [internal-to-Apollo context](https://apollograph.slack.com/archives/C04LHRSTLDD/p1692300519065359?thread_ts=1692221675.357379&cid=C04LHRSTLDD) if the change isn't self-explanatory.)

garypen commented on apollo-router/src/services/layers/persisted_queries/manifest_poller.rs:
```diff
@@ -18,102 +21,322 @@ use crate::uplink::persisted_queries_manifest_stream::PersistedQueriesManifestCh
+    },
+}
+
+/// Describes what the router should do for a given request: allow it, deny it
+/// with an error, or allow it but log the operation as unknown.
```
Nit. The comment appears to have drifted out of date with the implementation. For this kind of reason, I prefer comments like this:

```
/// How should we treat Freeform GraphQL
pub(crate) enum FreeformGraphQLAction {
    /// Allow it
    Allow,
    /// Deny it, with an error
    Deny,
    /// Allow it, but log the operation as unknown
    AllowAndLog,
    /// Deny it, with an error and log the operation as unknown <- Presumably...
    DenyAndLog,
}
```

abernix replied:
A durable and defensive suggetion!  Love it.

glasser replied:
Thanks @BrynCooke for merging! Should I make a follow-up with this comment improvement?

garypen replied:
If you don't mind.

garypen reviewed (commented):
Looks like a nice set of improvements.
